src/gfx.cpp

```diff
@@ -1445,20 +1438,75 @@ void RedrawScreenRect(int left, int top, int right, int bottom)
 	VideoDriver::GetInstance()->MakeDirty(left, top, right - left, bottom - top);
 }
 
+static std::vector<Rect> _dirty_viewport_occlusions;
+static const Viewport *_dirty_viewport;
+static NWidgetDisplay _dirty_viewport_disp_flags;
+
+static void DrawDirtyViewport(uint occlusion, int left, int top, int right, int bottom)
+{
+	for(; occlusion < _dirty_viewport_occlusions.size(); occlusion++) {
+		const Rect &occ = _dirty_viewport_occlusions[occlusion];
+		if (right <= occ.left || bottom <= occ.top || left >= occ.right || top >= occ.bottom) {
+			/* No intersection. */
+			continue;
+		}
+
+		int middle;
+		if (left < (middle = occ.left)) {
+			/* Occlusion intersects with dirty rect; split at left edge. */
+			DrawDirtyViewport(occlusion + 1, left, top, middle, bottom);
+			DrawDirtyViewport(occlusion, middle, top, right, bottom);
+			return;
+		}
+
+		if (right > (middle = occ.right)) {
+			/* Occlusion intersects with dirty rect; split at right edge. */
+			DrawDirtyViewport(occlusion, left, top, middle, bottom);
+			DrawDirtyViewport(occlusion + 1, middle, top, right, bottom);
+			return;
+		}
+
+		if (top < (middle = occ.top)) {
+			/* Occlusion intersects with dirty rect; split at top edge. */
+			DrawDirtyViewport(occlusion + 1, left, top, right, middle);
+			DrawDirtyViewport(occlusion, left, middle, right, bottom);
+			return;
+		}
+
+		if (bottom > (middle = occ.bottom)) {
+			/* Occlusion intersects with dirty rect; split at bottom edge. */
+			DrawDirtyViewport(occlusion, left, top, right, middle);
+			DrawDirtyViewport(occlusion + 1, left, middle, right, bottom);
+			return;
+		}
+
+		/* Dirty rect is completely occluded. Nothing more to do. */
+		return;
+	}
+
+	if (_game_mode == GM_MENU) {
+		RedrawScreenRect(left, top, right, bottom);
+	} else {
+		extern void ViewportDrawChk(const Viewport *vp, int left, int top, int right, int bottom);
+		ViewportDrawChk(_dirty_viewport, left, top, right, bottom);
+		if (_dirty_viewport_disp_flags & (ND_SHADE_GREY | ND_SHADE_DIMMED)) {
+			GfxFillRect(left, top, right, bottom,
+					(_dirty_viewport_disp_flags & ND_SHADE_DIMMED) ? PALETTE_TO_TRANSPARENT : PALETTE_NEWSPAPER, FILLRECT_RECOLOUR);
+		}
+		VideoDriver::GetInstance()->MakeDirty(left, top, right - left, bottom - top);
+	}
+}
+
 /**
  * Repaints the rectangle blocks which are marked as 'dirty'.
  *
- * @see AddDirtyBlock
+ * @see SetDirtyBlocks
  *
  * @ingroup dirty
  */
 void DrawDirtyBlocks()
```

Comment:
At some point, I will be tearing apart both `DrawDirtyBlocks` and `DrawDirtyViewport` and documenting and/or refactoring them until I have a clear understanding of what they do.

When I started https://github.com/OpenTTD/OpenTTD/pull/8214, I quickly decided that it was too much effort to save the existing code and just ended up rewriting it. However, the algorithm has changed significantly since then, so the only option is to carefully refactor it and document it as I go.

This is significant work, so it will be sensible to address the other remaining issues in this pull request and then tackle these as a supplemental PR.